src/lib.rs

```diff
@@ -31,124 +32,221 @@ use core::{
 /// Every block of data is processed in chunks of `BYTES` bytes.
 pub const BYTES: usize = 64;
 
+enum ScannerState {
+    PreAlign(usize),
+    Simd(u64),
+    Tail,
+    End,
+}
+
 /// An iterator for searching a given pattern in data
 pub struct Scanner<'pattern, 'data: 'cursor, 'cursor> {
     pattern: &'pattern Pattern,
     data: &'data [u8],
     cursor: &'cursor [u8],
-    position: usize,
-    buffer: Buffer,
+    state: ScannerState,
 }
 
 impl<'pattern, 'data: 'cursor, 'cursor> Scanner<'pattern, 'data, 'cursor> {
     /// Create an iterator, also see [`Pattern::matches`]
     #[must_use]
     #[inline]
     pub fn new(pattern: &'pattern Pattern, data: &'data [u8]) -> Scanner<'pattern, 'data, 'cursor> {
+        let align = data.as_ptr().align_offset(BYTES);
         Scanner {
             pattern,
             data,
             cursor: data,
-            buffer: Buffer::new(),
-            position: 0,
+            state: if align < BYTES {
+                // by contract, align_offset may return usize::MAX
+                ScannerState::PreAlign(min(data.len(), align)) // if data < BYTES, it's all prefix
+            } else {
+                ScannerState::Simd(0)
+            },
         }
     }
 }
 
+/// Match `pattern` against the start of `data` (without SIMD)
+#[inline(always)]
+fn plain_match(pattern: &Pattern, data: &[u8]) -> bool {
+    pattern.bytes.as_array()[0..pattern.length]
+        .iter()
+        .zip(pattern.mask.to_array()[0..pattern.length].iter())
+        .zip(data[0..pattern.length].iter())
+        .all(|((pat, mask), val)| (!*mask) || *pat == *val)
+}
+
+/// Find the offset of `cursor` into `data`.
+///
+/// # Safety
+/// Assumes that `cursor` is derived from `data` as in `data[K..]`
+#[inline(always)]
+fn cursor_offset(cursor: &&[u8], data: &[u8]) -> usize {
+    (unsafe { cursor.as_ptr().offset_from(data.as_ptr()) } as usize)
+}
```

Comment:
Would it be more correct to mark the function as `unsafe`, or move it into an `impl` on `Scanner`?
The latter is problematic since the call sites use split borrows rather than borrowing all of `Scanner`

Might not need to do anything since it's a private function, I guess.

Reply:
Just mark it unsafe :slightly_smiling_face: